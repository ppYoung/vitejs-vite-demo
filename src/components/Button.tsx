import { defineComponent, useSlots } from 'vue';

export default defineComponent({
  props: {
    type: {
      type: String,
      default: 'default',
    },
    loadingText: {
      type: String,
      default: '',
    },
    loading: {
      type: Boolean,
      default: false,
    },
    disabled: {
      type: Boolean,
      default: false,
    },
    size: {
      type: String,
      default: 'normal',
    },
  },
  emits: ['click'],
  setup(props, { emit }) {
    const slots = useSlots();

    interface KVStr {
      [k: string]: string | string[];
    }

    const TYPE_CLASS_MAP: KVStr = {
      default: ['bg-white', 'active:bg-gray-200', 'border', 'border-gray-200'],
      primary: ['bg-blue-500', 'active:bg-blue-400'],
      success: ['bg-emerald-500', 'active:bg-emerald-400'],
      warning: ['bg-amber-500', 'active:bg-amber-500'],
      danger: ['bg-red-500', 'active:bg-red-500'],
    };

    const SIZE_CLASS_MAP: KVStr = {
      small: ['min-w-12', 'min-h-6', 'tracking-normal', 'text-sm', 'px-1.5'],
      normal: ['min-w-16', 'min-h-8', 'tracking-wide', 'text-base', 'px-2'],
      large: ['min-w-20', 'min-h-10', 'tracking-wider', 'text-large', 'px-4'],
    };

    const classes = [
      'd-button',
      props.type === 'default' ? 'color-dark' : 'color-light',
      'inline-flex',
      'justify-center',
      'items-center',
      'rounded-md',
      props.disabled ? '' : 'hover:opacity-90',
      'disabled:opacity-50',
      'disabled:cursor-not-allowed',
      TYPE_CLASS_MAP[props.type],
      SIZE_CLASS_MAP[props.size],
    ];

    const renderIcon = () => {
      if (props.loading) {
        return slots.loadingIcon ? (
          slots.loadingIcon()
        ) : (
          <i class={['inline-block', 'i-eos-icons:loading', 'mr-0.5']} />
        );
      }

      if (slots.icon) {
        return slots.icon();
      }
    };

    const onClick = () => {
      if (props.loading || props.disabled) return;
      emit('click');
    };

    return () => (
      <button class={classes} onClick={onClick} disabled={props.disabled}>
        {renderIcon()}
        {slots?.default?.()}
      </button>
    );
  },
});
