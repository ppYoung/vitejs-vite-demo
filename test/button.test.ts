import { shallowMount } from '@vue/test-utils';
import { expect, test } from 'vitest';

import Button from '../src/components/Button';

test('should mount component correctly', () => {
  const wrapper = shallowMount(Button, {
    props: {
      type: 'primary',
    },
  });

  expect(wrapper.html()).toMatchSnapshot();
});

test('should render button text correctly', () => {
  const wrapper = shallowMount(Button, {
    slots: {
      default: 'My Button',
    },
  });

  expect(wrapper.text()).toBe('My Button');
  expect(wrapper.html()).toMatchSnapshot();
});

test('should render primary correctly', () => {
  const wrapper = shallowMount(Button, {
    props: {
      type: 'primary',
    },
  });

  expect(wrapper.classes()).includes('bg-blue-500');
  expect(wrapper.html()).toMatchSnapshot();
});

test('should render disabled state correctly', () => {
  const wrapper = shallowMount(Button, {
    props: {
      disabled: true,
    },
  });

  expect(wrapper.classes()).includes('disabled:cursor-not-allowed');
  expect(wrapper.html()).toMatchSnapshot();
});

test('should render loading state correctly', () => {
  const wrapper = shallowMount(Button, {
    props: {
      loading: true,
    },
  });

  expect(wrapper.classes()).includes('disabled:cursor-not-allowed');
  expect(wrapper.find('i').classes()).includes('i-eos-icons:loading');
  expect(wrapper.html()).toMatchSnapshot();
});

test('should emit click event correctly', async () => {
  const wrapper = shallowMount(Button, {});

  await wrapper.trigger('click');
  expect(wrapper.emitted('click')).toHaveLength(1);
  expect(wrapper.html()).toMatchSnapshot();
});

test('should emit loading click event correctly', async () => {
  const wrapper = shallowMount(Button, {
    props: {
      loading: true,
    },
  });

  await wrapper.trigger('click');
  expect(wrapper.emitted('click')).toBeUndefined();
  expect(wrapper.html()).toMatchSnapshot();
});

test('should emit disabled click event correctly', async () => {
  const wrapper = shallowMount(Button, {
    props: {
      disabled: true,
    },
  });

  await wrapper.trigger('click');
  expect(wrapper.emitted('click')).toBeUndefined();
  expect(wrapper.html()).toMatchSnapshot();
});
