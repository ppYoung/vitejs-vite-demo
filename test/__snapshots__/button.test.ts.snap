// Vitest Snapshot v1

exports[`should emit click event correctly 1`] = `
"<button class=\\"d-button color-dark inline-flex justify-center items-center rounded-md hover:opacity-90 disabled:opacity-50 disabled:cursor-not-allowed bg-white active:bg-gray-200 border border-gray-200 min-w-16 min-h-8 tracking-wide text-base px-2\\">
  <!---->
  <!---->
</button>"
`;

exports[`should emit disabled click event correctly 1`] = `
"<button class=\\"d-button color-dark inline-flex justify-center items-center rounded-md disabled:opacity-50 disabled:cursor-not-allowed bg-white active:bg-gray-200 border border-gray-200 min-w-16 min-h-8 tracking-wide text-base px-2\\" disabled=\\"\\">
  <!---->
  <!---->
</button>"
`;

exports[`should emit loading click event correctly 1`] = `
"<button class=\\"d-button color-dark inline-flex justify-center items-center rounded-md hover:opacity-90 disabled:opacity-50 disabled:cursor-not-allowed bg-white active:bg-gray-200 border border-gray-200 min-w-16 min-h-8 tracking-wide text-base px-2\\"><i class=\\"inline-block i-eos-icons:loading mr-0.5\\"></i>
  <!---->
</button>"
`;

exports[`should mount component correctly 1`] = `
"<button class=\\"d-button color-light inline-flex justify-center items-center rounded-md hover:opacity-90 disabled:opacity-50 disabled:cursor-not-allowed bg-blue-500 active:bg-blue-400 min-w-16 min-h-8 tracking-wide text-base px-2\\">
  <!---->
  <!---->
</button>"
`;

exports[`should render button text correctly 1`] = `
"<button class=\\"d-button color-dark inline-flex justify-center items-center rounded-md hover:opacity-90 disabled:opacity-50 disabled:cursor-not-allowed bg-white active:bg-gray-200 border border-gray-200 min-w-16 min-h-8 tracking-wide text-base px-2\\">
  <!---->My Button
</button>"
`;

exports[`should render disabled state correctly 1`] = `
"<button class=\\"d-button color-dark inline-flex justify-center items-center rounded-md disabled:opacity-50 disabled:cursor-not-allowed bg-white active:bg-gray-200 border border-gray-200 min-w-16 min-h-8 tracking-wide text-base px-2\\" disabled=\\"\\">
  <!---->
  <!---->
</button>"
`;

exports[`should render loading state correctly 1`] = `
"<button class=\\"d-button color-dark inline-flex justify-center items-center rounded-md hover:opacity-90 disabled:opacity-50 disabled:cursor-not-allowed bg-white active:bg-gray-200 border border-gray-200 min-w-16 min-h-8 tracking-wide text-base px-2\\"><i class=\\"inline-block i-eos-icons:loading mr-0.5\\"></i>
  <!---->
</button>"
`;

exports[`should render primary correctly 1`] = `
"<button class=\\"d-button color-light inline-flex justify-center items-center rounded-md hover:opacity-90 disabled:opacity-50 disabled:cursor-not-allowed bg-blue-500 active:bg-blue-400 min-w-16 min-h-8 tracking-wide text-base px-2\\">
  <!---->
  <!---->
</button>"
`;
